### ISSUE TITLE
JNA required on windows

### ISSUE THREAD

**Issue by andrewscode** [dnsjava/dnsjava#112]
JNA is currently required on windows.    Although the code seems to imply it checks for JNA:
```
 public WindowsResolverConfigProvider() {
    if (System.getProperty("os.name").contains("Windows")) {
      try {
        this.inner = new WindowsResolverConfigProvider.InnerWindowsResolverConfigProvider();
      } catch (NoClassDefFoundError var2) {
        log.debug("JNA not available");
      }
    }

  }
```
Just creating the WindowsResolverConfigProvider.InnerWindowsResolverConfigProvider will not cause a NoClassDefFoundError exception.

But will be thrown later on by the initialize() method causing the entire ResolverConfig to fail

```
Caused by: java.lang.NoClassDefFoundError: com/sun/jna/Memory
	at org.xbill.DNS.config.WindowsResolverConfigProvider$InnerWindowsResolverConfigProvider.initialize(WindowsResolverConfigProvider.java:53)
	at org.xbill.DNS.config.WindowsResolverConfigProvider.initialize(WindowsResolverConfigProvider.java:108)
	at org.xbill.DNS.ResolverConfig.<init>(ResolverConfig.java:84)
	at org.xbill.DNS.ResolverConfig.refresh(ResolverConfig.java:74)
	at org.xbill.DNS.ResolverConfig.<clinit>(ResolverConfig.java:59)
```

**Comment by ibauersachs** [comment]
I can't verify this with the following minimal example (Sample code, console output, Maven config). Can you please provide some more context? Can you provide the content of your `os.name` property? Which JDK are you running this on?

```java
import org.xbill.DNS.*;
public class Test {
  public static void main(String[] args) throws TextParseException {
    ResolverConfig.refresh();
    Lookup lookup = new Lookup("example.com");
    for (Record r : lookup.run()) {
      System.out.println(r);
    }
  }
}
```

```
[main] DEBUG org.xbill.DNS.config.WindowsResolverConfigProvider - JNA not available
[main] DEBUG org.xbill.DNS.config.JndiContextResolverConfigProvider$InnerJndiContextResolverConfigProvider - Added /62.2.17.61:53 to nameservers
[main] DEBUG org.xbill.DNS.config.JndiContextResolverConfigProvider$InnerJndiContextResolverConfigProvider - Added /62.2.24.158:53 to nameservers
[main] DEBUG org.xbill.DNS.config.JndiContextResolverConfigProvider$InnerJndiContextResolverConfigProvider - Added /62.2.17.60:53 to nameservers
[main] DEBUG org.xbill.DNS.config.JndiContextResolverConfigProvider$InnerJndiContextResolverConfigProvider - Added /62.2.24.162:53 to nameservers
[main] DEBUG org.xbill.DNS.Lookup - lookup example.com. A, cache answer: unknown
[main] DEBUG org.xbill.DNS.ExtendedResolver - Sending example.com./A, id=41041 to resolver 0 (SimpleResolver [/62.2.17.60:53]), attempt 1 of 3
[main] DEBUG org.xbill.DNS.SimpleResolver - Sending example.com./A, id=41041 to udp/62.2.17.60:53
[main] DEBUG org.xbill.DNS.Client - Starting dnsjava NIO selector thread
[main] DEBUG org.xbill.DNS.Cache - caching successful for example.com.
[main] DEBUG org.xbill.DNS.Lookup - Queried example.com. A: successful
[dnsjava NIO selector] DEBUG org.xbill.DNS.Client - dnsjava NIO selector thread stopped
example.com.		86210	IN	A	93.184.216.34
```

```xml
<project xmlns="http://maven.apache.org/POM/4.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <artifactId>sample</artifactId>
  <properties>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
  </properties>
  <dependencies>
    <dependency>
      <groupId>dnsjava</groupId>
      <artifactId>dnsjava</artifactId>
      <version>3.1.0</version>
    </dependency>
    <dependency>
      <groupId>org.slf4j</groupId>
      <artifactId>slf4j-simple</artifactId>
      <version>1.7.30</version>
    </dependency>
  </dependencies>
  <build>
    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>3.8.1</version>
        <configuration>
          <source>1.8</source>
          <target>1.8</target>
        </configuration>
      </plugin>
    </plugins>
  </build>
</project>
```

**Comment by andrewscode** [comment]
os.name = Windows 10

**Comment by andrewscode** [comment]
Seems to be fine on it's own but not under spring boot.

On a side note,  there doesn't seem to be a way of initializing the ResolverConfigProviders without initializing the ResolverConfig.refresh() being called first on it's existing list.

We always pass in a resolver for all Lookup.   So we don't want the default resolvers to be initialized.

**Comment by ibauersachs** [comment]
> Seems to be fine on it's own but not under spring boot.

I'm not familiar with Spring Boot. Any idea what it's doing different?

> On a side note, there doesn't seem to be a way of initializing the ResolverConfigProviders without initializing the ResolverConfig.refresh() being called first on it's existing list.

This is/was intentional to avoid the need of calling refresh first, and I need to keep that behavior for backwards compatibility. I could add a property that skips the initialization. Would that help you?
 
> We always pass in a resolver for all Lookup. So we don't want the default resolvers to be initialized.

**Comment by andrewscode** [comment]
> > Seems to be fine on it's own but not under spring boot.
> 
> I'm not familiar with Spring Boot. Any idea what it's doing different?
> 

I'm not entirely certain.   If you provide that property to skip initialization that'd definitely help my problem.

Thanks

**Comment by ibauersachs** [comment]
The property `dnsjava.configprovider.skipinit` is already in master, see the linked commit. You can expect a release in a couple of days or so.

**Comment by andrewscode** [comment]
I've spent a bit more time looking at this.  

When running standalone,   the NoClassDefFoundError is thrown for the class  com.sun.jna.Pointer.   This doesn't happen with SpringBoot.  But the real question is why does it get thrown without spring boot?      

The WindowsResolverConfigProvider creates a new InnerWindowsResolverConfigProvider.    That has no constructor and although the parent BaseResolverConfigProvider has some local members,  nothing constructs a com.sun.jna.Pointer object.    I wouldn't expect a NoClassDefFoundError thrown when constructing that Inner class.

I think what would help is simply calling Class.forName("com.sun.jna.Pointer", false, this.getClass().getClassLoader()) in the constructor of WindowsResolverConfigProvider.   If a ClassNotFoundException is thrown, then you know jna is not on the classpath.    

`
 try {
      Class.forName("com.sun.jna.Pointer", false, this.getClass().getClassLoader());
      Class.forName("com.sun.jna.platform.win32.Win32Exception", false, this.getClass().getClassLoader());

    } catch (ClassNotFoundException e) {
      log.debug("JNA not available");
    }
`

I test Win32Exception as well as that is apart of the jna-platform (whereas Pointer is in the jna library)

**Comment by ibauersachs** [comment]
I did some reading, and the different behavior is allowed according to [JLS §12.3](https://docs.oracle.com/javase/specs/jls/se8/html/jls-12.html#jls-12.3). Class loading consists of verification, preparation, resolution, initialization. Resolution can _optionally_ be lazy, i.e. references to other classes like Pointer or Win32Exception can be resolved on demand; _after_ initialization (= static and instance constructor(s)). Spring Boot seems to take adavantage of that option.

I went for a slightly different approach: reference the two required classes in the static constructor. To pass initialization, they _must_ be resolved. This is equivalent to Class.forName, the only difference is avoiding reflection (+ the log statement).
